`Testing.ZopeTestCase.FunctionalTestCase.publish` can not be used with %-encoded path or unicode paths.

## BUG/PROBLEM REPORT (OR OTHER COMMON ISSUE)

`Testing.ZopeTestCase.FunctionalTestCase.publish` can not be used with %-encoded or unicode paths.

### What I did:

I wrote a test for a Zope application, in this application some request paths are using non-ascii characters, I was not able to use `ZopeTestCase.FunctionalTestCase.publish` to test such requests.

A minimal reproduction would be something like this:

```py
# coding: utf-8
from Testing import ZopeTestCase


class MinimalReproductionTest(ZopeTestCase.FunctionalTestCase):
    def afterSetUp(self):
        self.folder_path = '/' + self.folder.absolute_url(1)
        self.folder.addDTMLMethod('täst', file=b'test')

    def testPublishDocumentNonAscii(self):
        response = self.publish(self.folder_path + '/täst')
        #   File "Zope/src/ZPublisher/WSGIPublisher.py", line 336, in publish_module
        #     path_info = path_info.decode('utf-8')
        # UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe4 in position 17: invalid continuation byte
        self.assertEqual(response.getStatus(), 200)

    def testPublishDocumentNonAsciiUrlEncoded(self):
        response = self.publish(self.folder_path + '/t%C3%A4st')
        self.assertEqual(response.getStatus(), 200)
        # AssertionError: 404 != 200
```

### What I expect to happen:

Tests pass

### What actually happened:

Both tests failed

### What version of Python and Zope/Addons I am using:

This is on Zope master on python 3.7.7

`testPublishDocumentNonAsciiUrlEncoded` never worked,
`testPublishDocumentNonAscii` was OK on Zope 4 and python 2


## Some more details from investigating

`publish(path, ...)` acts like a WSGI server, it prepares the environment and then calls `publish_module` (which is Zope's WSGI application), but because it does not do encode the `PATH_INFO` in environment like a WSGI server is supposed to do.

wsgiref does [this](https://github.com/python/cpython/blob/dd53b79de0ea98af6a11481217a961daef4e9774/Lib/wsgiref/simple_server.py#L80-L86):

```py
        if '?' in self.path:
            path,query = self.path.split('?',1)
        else:
            path,query = self.path,''

        env['PATH_INFO'] = urllib.parse.unquote(path, 'iso-8859-1')
        env['QUERY_STRING'] = query
```

waitress does [something similar](https://github.com/Pylons/waitress/blob/73fe701cba0b37dc6099858b064cb1e755e83e9e/src/waitress/parser.py#L376).

I believe we can change `ZopeTestCase.FunctionalTestCase.publish` to also change url-decode the path decode the bytes as iso-8859-1. I will send a pull request soon.

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -12,6 +12,9 @@ https://github.com/zopefoundation/Zope/blob/4.x/CHANGES.rst
 
 - Update to newest compatible versions of dependencies.
 
+- Change functional testing utilities to support percent encoded and unicode
+  paths (`#1058 <https://github.com/zopefoundation/Zope/issues/1058>`_).
+
 
 5.6 (2022-09-09)
 ----------------

--- a/src/Testing/ZopeTestCase/functional.py
+++ b/src/Testing/ZopeTestCase/functional.py
@@ -17,6 +17,7 @@ After Marius Gedminas' functional.py module for Zope3.
 
 import sys
 from functools import partial
+from urllib.parse import unquote_to_bytes
 
 import transaction
 from Testing.ZopeTestCase import interfaces
@@ -80,13 +81,11 @@ class Functional(sandbox.Sandboxed):
         env['SERVER_PROTOCOL'] = 'HTTP/1.1'
         env['REQUEST_METHOD'] = request_method
 
-        p = path.split('?')
-        if len(p) == 1:
-            env['PATH_INFO'] = p[0]
-        elif len(p) == 2:
-            [env['PATH_INFO'], env['QUERY_STRING']] = p
-        else:
-            raise TypeError('')
+        query = ''
+        if '?' in path:
+            path, query = path.split("?", 1)
+        env['PATH_INFO'] = unquote_to_bytes(path).decode('latin-1')
+        env['QUERY_STRING'] = query
 
         if basic:
             env['HTTP_AUTHORIZATION'] = basic_auth_encode(basic)
@@ -97,12 +96,16 @@ class Functional(sandbox.Sandboxed):
 
         if stdin is None:
             stdin = BytesIO()
+        env['wsgi.input'] = stdin
 
         outstream = BytesIO()
         response = WSGIResponse(stdout=outstream, stderr=sys.stderr)
-        request = Request(stdin, env, response)
-        for k, v in extra.items():
-            request[k] = v
+
+        def request_factory(*args):
+            request = Request(*args)
+            for k, v in extra.items():
+                request[k] = v
+            return request
 
         wsgi_headers = BytesIO()
 
@@ -120,7 +123,10 @@ class Functional(sandbox.Sandboxed):
             wsgi_headers.write(headers)
             wsgi_headers.write(b'\r\n\r\n')
 
-        publish = partial(publish_module, _request=request, _response=response)
+        publish = partial(
+            publish_module,
+            _request_factory=request_factory,
+            _response=response)
         if handle_errors:
             publish = HTTPExceptionHandler(publish)
 

--- a/src/Testing/ZopeTestCase/testFunctional.py
+++ b/src/Testing/ZopeTestCase/testFunctional.py
@@ -60,6 +60,9 @@ class TestFunctional(ZopeTestCase.FunctionalTestCase):
         # A method changing the title property of an object
         self.folder.addDTMLMethod('change_title', file=CHANGE_TITLE_DTML)
 
+        # A method with a non-ascii path
+        self.folder.addDTMLMethod('täst', file=b'test')
+
     def testPublishFolder(self):
         response = self.publish(self.folder_path)
         self.assertEqual(response.getStatus(), 200)
@@ -69,6 +72,16 @@ class TestFunctional(ZopeTestCase.FunctionalTestCase):
         response = self.publish(self.folder_path + '/index_html')
         self.assertEqual(response.getStatus(), 200)
         self.assertEqual(response.getBody(), b'index')
+
+    def testPublishDocumentNonAscii(self):
+        response = self.publish(self.folder_path + '/täst')
+        self.assertEqual(response.getStatus(), 200)
+        self.assertEqual(response.getBody(), b'test')
+
+    def testPublishDocumentNonAsciiUrlEncoded(self):
+        response = self.publish(self.folder_path + '/t%C3%A4st')
+        self.assertEqual(response.getStatus(), 200)
+        self.assertEqual(response.getBody(), b'test')
 
     def testUnauthorized(self):
         response = self.publish(self.folder_path + '/secret_html')

--- a/src/Testing/ZopeTestCase/zopedoctest/functional.py
+++ b/src/Testing/ZopeTestCase/zopedoctest/functional.py
@@ -131,7 +131,7 @@ def http(request_string, handle_errors=True):
 
     This is used for HTTP doc tests.
     """
-    from urllib.parse import unquote
+    from urllib.parse import unquote_to_bytes
 
     from ZPublisher.HTTPRequest import WSGIRequest as Request
     from ZPublisher.HTTPResponse import WSGIResponse
@@ -148,7 +148,6 @@ def http(request_string, handle_errors=True):
     command_line = request_string[:newline].rstrip()
     request_string = request_string[newline + 1:]
     method, path, protocol = command_line.split()
-    path = unquote(path)
 
     env = {
         'HTTP_HOST': 'localhost',
@@ -157,13 +156,11 @@ def http(request_string, handle_errors=True):
         'SERVER_PROTOCOL': protocol,
     }
 
-    p = path.split('?', 1)
-    if len(p) == 1:
-        env['PATH_INFO'] = p[0]
-    elif len(p) == 2:
-        [env['PATH_INFO'], env['QUERY_STRING']] = p
-    else:
-        raise TypeError('')
+    query = ''
+    if '?' in path:
+        path, query = path.split("?", 1)
+    env['PATH_INFO'] = unquote_to_bytes(path).decode('latin-1')
+    env['QUERY_STRING'] = query
 
     header_output = HTTPHeaderOutput(
         protocol, ('x-content-type-warning', 'x-powered-by'))
@@ -196,7 +193,6 @@ def http(request_string, handle_errors=True):
 
     outstream = BytesIO()
     response = WSGIResponse(stdout=outstream, stderr=sys.stderr)
-    request = Request(instream, env, response)
 
     env['wsgi.input'] = instream
     wsgi_headers = BytesIO()
@@ -209,7 +205,10 @@ def http(request_string, handle_errors=True):
         wsgi_headers.write(headers)
         wsgi_headers.write(b'\r\n\r\n')
 
-    publish = partial(publish_module, _request=request, _response=response)
+    publish = partial(
+        publish_module,
+        _request_factory=Request,
+        _response=response)
     if handle_errors:
         publish = HTTPExceptionHandler(publish)
 

